okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt

```diff
@@ -445,93 +462,90 @@ class FakeFileSystem(
       check(!closed) { "closed" }
-      return size
+      return file.data.size.toLong()
     }
 
-    override fun seek(position: Long) {
+    override fun protectedRead(
+      fileOffset: Long,
+      array: ByteArray,
+      arrayOffset: Int,
+      byteCount: Int
+    ): Int {
       check(!closed) { "closed" }
-
-      buffer.clear()
-      when {
-        position <= size -> {
-          buffer.write(file.data)
-          size = file.data.size.toLong()
-          buffer.skip(position)
-        }
-        else -> {
-          size = position
-        }
+      check(readAccess) { "not opened for read" }
+      checkOffsetAndCount(array.size.toLong(), arrayOffset.toLong(), byteCount.toLong())
+
+      val fileOffsetInt = fileOffset.toInt()
+      val toCopy = minOf(file.data.size - fileOffsetInt, byteCount)
+      if (toCopy <= 0) return -1
+      for (i in 0 until toCopy) {
+        array[i + arrayOffset] = file.data[i + fileOffsetInt]
       }
+      return toCopy
     }
 
-    override fun close() {
-      if (closed) return
-      closed = true
-      openFiles -= openFile
-    }
-
-    override fun toString() = "source(${openFile.canonicalPath})"
-  }
-
-  /** Writes data to [path]. */
-  private inner class FakeFileSink(
-    private val openFile: OpenFile,
-    private val file: File
-  ) : Sink {
-    val buffer = Buffer()
-    private var closed = false
-
-    override fun write(source: Buffer, byteCount: Long) {
+    override fun protectedWrite(
+      fileOffset: Long,
+      array: ByteArray,
+      arrayOffset: Int,
+      byteCount: Int
+    ) {
       check(!closed) { "closed" }
-      buffer.write(source, byteCount)
-    }
+      check(writeAccess) { "not opened for write" }
+      checkOffsetAndCount(array.size.toLong(), arrayOffset.toLong(), byteCount.toLong())
 
-    override fun flush() {
-      check(!closed) { "closed" }
+      val buffer = Buffer()
+      buffer.write(file.data, 0, minOf(fileOffset.toInt(), file.data.size))
+      while (buffer.size < fileOffset) {
+        buffer.writeByte(0)
+      }
+      buffer.write(array, arrayOffset, byteCount)
+      if (buffer.size < file.data.size) {
+        buffer.write(file.data, buffer.size.toInt(), file.data.size - buffer.size.toInt())
+      }
       file.data = buffer.snapshot()
       file.access(now = clock.now(), modified = true)
     }
 
-    override fun timeout() = Timeout.NONE
+    override fun protectedFlush() {
+      check(!closed) { "closed" }
+      check(writeAccess) { "not opened for write" }
+    }
 
-    override fun close() {
+    override fun protectedClose() {
       if (closed) return
       closed = true
-      file.data = buffer.snapshot()
-      file.access(now = clock.now(), modified = true)
+      file.access(now = clock.now(), modified = writeAccess)
       openFiles -= openFile
     }
 
-    override fun toString() = "sink(${openFile.canonicalPath})"
+    override fun toString() = "FileHandler(${openFile.canonicalPath})"
```

Comment:
Why not FakeFileHandle?

Reply:
It’s not a public type name, so it’s a bit weird to show up in toString(). Anyone who wants that information can get it from `.getClass()`.